lib/blocks.js

```diff
@@ -42,29 +46,68 @@ api.getNextBlockInfo = (ledgerNode, callback) => {
   // does not have that information. There has presently been no mechanism
   // devised for switching consensus methods between hashgraph-like blocks
   // and typical blockchains with block heights.
-  ledgerNode.storage.blocks.getLatestSummary((err, block) => {
+  const ledgerNodeId = ledgerNode.id;
+  async.auto({
+    latestSummary: callback => ledgerNode.storage.blocks.getLatestSummary(
+      (err, block) => {
+        const latestBlockHeight = _.get(block, 'eventBlock.block.blockHeight');
+        if(latestBlockHeight === undefined) {
+          return callback(new BedrockError(
+            'blockHeight is missing from latest block.', 'NotFoundError',
+            {block}));
+        }
+        callback(null, block);
+      }),
+    cachedBlockHeight: callback => {
+      const blockHeightKey = _cacheKey.blockHeight(ledgerNodeId);
+      cache.client.get(blockHeightKey, (err, result) => {
+        if(err) {
+          return callback(err);
+        }
+        if(result === null) {
+          return callback(new BedrockError(
+            'Block height key is missing for the cache.', 'InvalidStateError',
+            {ledgerNodeId}));
+        }
+        callback(null, parseInt(result, 10));
+      });
+    },
+    consistencyCheck: [
+      'cachedBlockHeight', 'latestSummary', (results, callback) => {
+        const {cachedBlockHeight, latestSummary} = results;
+        const {blockHeight} = latestSummary.eventBlock.block;
+        if(blockHeight === cachedBlockHeight) {
+          // success
+          return callback();
+        }
+        // this should never happen
+        if((blockHeight - cachedBlockHeight) !== 1) {
+          return callback(new BedrockError(
+            'Cache is behind by more than one block.', 'InvalidStateError',
+            {blockHeight, cachedBlockHeight, ledgerNodeId}));
+        }
+        // inconsistency needs to be repaired
+        _repairCache({blockHeight, ledgerNode}, callback);
```

Comment:
Why are we calling the repair code within `getLatest`? This call can be called from anywhere -- right? If so, is it necessary to do this cache repair when that happens (i.e. is the cache even used in that case?) How safe is it to have two processes make this call at once (i.e. outside of the consensus worker lock)?

Would it be better to just call the repair code once at the start of the consensus worker (or beginning of or start of loop of whatever agent is using the cache)?

Reply:
Relocated the consistency check.